File: AndroidRxJavaWithRetrofit2/app/src/main/java/com/blackbirds/shakil/androidrxjavawithretrofit2/MainActivity.java
package com.blackbirds.shakil.androidrxjavawithretrofit2;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;

import com.blackbirds.shakil.androidrxjavawithretrofit2.Adapter.PostListAdapter;
import com.blackbirds.shakil.androidrxjavawithretrofit2.Common.Common;
import com.blackbirds.shakil.androidrxjavawithretrofit2.Model.PostModel;
import com.blackbirds.shakil.androidrxjavawithretrofit2.Network.APIService;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {

    APIService apiService;
    RecyclerView recyclerViewPost;
    PostListAdapter adapter;
    CompositeDisposable compositeDisposable = new CompositeDisposable();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        apiService = Common.getAPIService();

        recyclerViewPost = findViewById(R.id.recyclerViewPost);
        recyclerViewPost.setLayoutManager(new LinearLayoutManager(this));

        fetchData();
    }

    private void fetchData() {
        compositeDisposable.add(apiService.getPostsList()
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(postModels -> {
            displayData(postModels);
        }));
    }

    private void displayData(List<PostModel> postModels) {
        adapter = new PostListAdapter(this, postModels);
        adapter.notifyDataSetChanged();
        recyclerViewPost.setAdapter(adapter);
    }

    @Override
    protected void onStop() {
        compositeDisposable.clear();
        super.onStop();
    }
}

File: KotlinRxJavaWithRetrofit2/app/src/main/java/com/blackbirds/shakil/kotlinrxjavawithretrofit2/MainActivity.kt
package com.blackbirds.shakil.kotlinrxjavawithretrofit2

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.blackbirds.shakil.kotlinrxjavawithretrofit2.Adapter.PostListAdapter
import com.blackbirds.shakil.kotlinrxjavawithretrofit2.Common.Common
import com.blackbirds.shakil.kotlinrxjavawithretrofit2.Model.PostModel
import com.blackbirds.shakil.kotlinrxjavawithretrofit2.Network.APIService
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class MainActivity : AppCompatActivity() {

    var apiService: APIService? = null
    var recyclerViewPost: RecyclerView? = null
    var adapter: PostListAdapter? = null
    var compositeDisposable = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        apiService = Common.aPIService

        recyclerViewPost = findViewById(R.id.recyclerViewPost)
        recyclerViewPost!!.setLayoutManager(LinearLayoutManager(this))

        fetchData()
    }

    private fun fetchData() {
        compositeDisposable.add(apiService!!.postsList()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { postModels -> displayData(postModels) })
    }

    private fun displayData(postModels: List<PostModel>) {
        adapter = PostListAdapter(this, postModels)
        adapter!!.notifyDataSetChanged()
        recyclerViewPost!!.adapter = adapter
    }

    override fun onStop() {
        compositeDisposable.clear()
        super.onStop()
    }
}